llvm/lib/Target/WebAssembly/WebAssemblyFastISel.cpp

```diff
@@ -337,6 +337,12 @@ bool WebAssemblyFastISel::computeAddress(const Value *Obj, Address &Addr) {
     break;
   }
   case Instruction::Add: {
+    // We should not fold operands into an offset when 'nuw' (no unsigned wrap)
+    // is not present, because the address calculation does not wrap.
+    if (auto *OFBinOp = dyn_cast<OverflowingBinaryOperator>(U))
+      if (!OFBinOp->hasNoUnsignedWrap())
```

Comment:
Do you have any sense of how often this happens, i.e. how often these address-calculation adds lack `nuw`? It would be a bummer if we bailed out of fast-isel for the majority of address-calculations (since bailing out forces the whole block to be re-selected, it could potentially regress compile time noticeably).

Reply:
I would assume this will be frontend-dependent. For C/C++ input, it should be rare because address arithmetic is all `inbounds` GEPs. For the input [we](https://github.com/llvm/llvm-project/issues/111018#issuecomment-2397415761) have, it will be very common.

Reply:
I made this pattern of code `assert(false)` and ran emscripten `core0` tests with it, and the only case this pattern was generated is in asan tests.

This part in the normal ISel is hit a lot more because many other instructions, including `getelementptr`s, are turned into `ISD:ADD` in ISel. But this is in the normal ISel and there's no "bailout". (What I told you offline was actually hitting this cases in the normal ISel. I added `assert(false)` in both here and the newly added FastISel checks. So this one is not very relevant anyway.)
https://github.com/llvm/llvm-project/blob/65688274b14f5d0e7dabbbaf9f3fd135646ef1d1/llvm/lib/Target/WebAssembly/WebAssemblyISelDAGToDAG.cpp#L328-L332

So I guess, as @SingleAccretion pointed out, we don't need to worry too much about -O0 compilation time at least in C/C++. But this may regress -O0 compilation time in C# (or other languages). 

Currently our FastISel's `selectLoad` and `selectStore` only handles the cases where its address can be represented within a single `Address` class:
https://github.com/llvm/llvm-project/blob/db4874c2fc0b34e8959ff0ac455bff5f54fbd52b/llvm/lib/Target/WebAssembly/WebAssemblyFastISel.cpp#L53-L110
And addressing modes are assumed to be either reg-based or frameindex-based. So we can handle cases like reg + label but not two different regs, which we need to handle here.

While it is not impossible to extend FastISel to handle the two-register cases, I guess that means we can handle a lot more cases that we are currently rejecting, which means even bigger overhaul. It's possible that we only handle the cases here, but we don't really know how frequent these cases compared to other cases we are already rejecting. Also given that this is a "Fast" ISel after all, I'm not sure whether a bigger extension of the pass is worthwhile.

How about landing this for now, because this is a correctness issue, and see if this is really regressing -O0 compilation times a lot? For that we'd appreciate feedback from C# devs (@SingleAccretion and @yowl).

Reply:
> and see if this is really regressing -O0 compilation times a lot

I can get some PIN (binary instrumentation) data on this change with our input - it should not take a lot of time (mainly building clang itself).

However, after looking at the code, I have a more basic question: why `return false` and not `break`? I don't think there is a need to support complex addressing modes with two registers, we can presumably just use the non-`nuw` `add` as the one and only base register (with a zero offset), which, if I am reading the code correctly, is what `break`ing will do.

Reply:
You are right! I was misunderstanding the "one reg" rule in FastISel. If FastISel cannot figure out how to make use of the offset, it assigns the whole expression to a single register, which then generates more instructions to make it happen (by adding more registers for its child expressions):
https://github.com/llvm/llvm-project/blob/c563fe5c71239a8938e1925fce4c19954564faf0/llvm/lib/Target/WebAssembly/WebAssemblyFastISel.cpp#L380

Anyway, changed them to `break`s, and renamed the test file and added `-fast-isel-abort=1` because we don't bail out of FastISel anymore. Thanks!